presto-main/src/main/java/com/facebook/presto/server/GracefulShutdownHandler.java

```
@@ -0,0 +1,126 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.google.common.util.concurrent.Uninterruptibles.sleepUninterruptibly;
+import static io.airlift.concurrent.Threads.threadsNamed;
+import static java.lang.Thread.currentThread;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadExecutor;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.MINUTES;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+public class GracefulShutdownHandler
+{
+    private static final Logger log = Logger.get(GracefulShutdownHandler.class);
+    private static final Duration GRACE_PERIOD = new Duration(2, MINUTES);
+    private static final Duration LIFECYCLE_STOP_TIMEOUT = new Duration(30, SECONDS);
+
+    private final ScheduledExecutorService shutdownHandler = newSingleThreadScheduledExecutor(threadsNamed("shutdown-handler-%s"));
+    private final ExecutorService lifeCycleStopper = newSingleThreadExecutor(threadsNamed("lifecycle-stopper-%s"));
+    private final TaskManager sqlTaskManager;
+    private final LifeCycleManager lifeCycleManager;
+    private final AtomicBoolean shutdownRequested = new AtomicBoolean(false);
+
+    @Inject
+    public GracefulShutdownHandler(TaskManager sqlTaskManager, LifeCycleManager lifeCycleManager)
+    {
+        this.sqlTaskManager = requireNonNull(sqlTaskManager, "sqlTaskManager is null");
+        this.lifeCycleManager = requireNonNull(lifeCycleManager, "lifeCycleManager is null");
+    }
+
+    public void requestShutdown()
+    {
+        log.info("Shutdown requested");
+
+        shutdownRequested.set(true);
```

@cberner Do we want to throw an exception if `shutdownRequested` is already true or just ignore the call silently?